packages/next-international/src/app/middleware/index.ts

```diff
@@ -1,42 +1,65 @@
 import { NextRequest, NextResponse } from 'next/server';
 
-const Response = NextResponse;
+import { LOCALE_COOKIE, LOCALE_HEADER } from '../../common/constants';
+import type { I18NMiddlewareConfig } from '../../types';
+
+const DEFAULT_STRATEGY = 'redirect';
 
 export function createI18nMiddleware<Locales extends readonly string[]>(
   locales: Locales,
   defaultLocale: Locales[number],
+  config?: I18NMiddlewareConfig,
 ) {
   return function I18nMiddleware(request: NextRequest) {
-    const pathname = request.nextUrl.pathname;
-    const pathnameIsMissingLocale = locales.every(
-      locale => !pathname.startsWith(`/${locale}/`) && pathname !== `/${locale}`,
-    );
+    const requestUrl = request.nextUrl.clone();
 
-    if (pathnameIsMissingLocale && pathname !== '/favicon.ico') {
-      let locale = request.cookies.get('Next-Locale')?.value;
+    const locale = localeFromRequest(request, locales, config?.contentNegotiator) ?? defaultLocale;
 
-      if (!locale) {
-        const acceptLanguage = request.headers.get('Accept-Language');
-        const acceptLanguageLocale = acceptLanguage?.split(',')?.[0]?.split('-')?.[0] ?? defaultLocale;
+    if (noLocalePrefix(locales, requestUrl.pathname)) {
+      const mappedUrl = requestUrl.clone();
+      mappedUrl.pathname = `/${locale}${mappedUrl.pathname}`;
 
-        locale = locales.includes(acceptLanguageLocale) ? acceptLanguageLocale : defaultLocale;
+      const strategy = config?.urlMappingStrategy ?? DEFAULT_STRATEGY;
+      if (strategy === 'redirect') {
+        const response = NextResponse.redirect(mappedUrl);
+        return addLocaleToResponse(response, locale);
+      } else if (strategy === 'rewrite') {
+        const response = NextResponse.rewrite(mappedUrl);
+        return addLocaleToResponse(response, locale);
+      } else {
+        throw new Error(`Invalid urlMappingStrategy: ${strategy}`);
       }
+    }
 
-      const response = Response.redirect(new URL(`/${locale}${pathname}`, request.url));
-
-      response.headers.set('X-Next-Locale', locale);
+    const response = NextResponse.next();
+    return addLocaleToResponse(response, locale);
+  };
+}
 
-      return response;
-    }
+function localeFromRequest(
+  request: NextRequest,
+  locales: readonly string[],
+  contentNegotiator?: (request: NextRequest, locales: readonly string[]) => string | null,
+) {
+  let locale = request.cookies.get(LOCALE_COOKIE)?.value ?? null;
+  if (!locale) {
+    locale = contentNegotiator ? contentNegotiator(request, locales) : negotiateAcceptLanguage(request);
```

Comment:
Do we need this `contentNegotiator` option for now? It adds a bit of complexity and I'm not sure if it's needed for now - in which cases don't you want to use the `Accept-Language` header to determine the user's locale?

Reply:
Agreed that it's not needed for now so removing.

GeoIP, domain, user preferences can all be used to factor into content negotiation. When I was testing this locally with my project, I wanted to use a library for this (like suggested in the [Next docs](https://nextjs.org/docs/app/building-your-application/routing/internationalization#routing-overview)) to support all the content negotiation edge cases like quality factors which the next-international middleware currently does not handle.

Reply:
Let's land this feature first, we can add the `contentNegotiator` option / any other useful thing to tweak later.